modules/product/src/ProductViewBuilder.php

```diff
@@ -0,0 +1,139 @@
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+class ProductViewBuilder extends EntityViewBuilder {
+
+  /**
+   * The product field variation renderer
+   *
+   * @var \Drupal\commerce_product\ProductVariationFieldRenderer
+   */
+  protected $productVariationFieldRenderer;
+
+  /**
+   * Constructs a new BlockViewBuilder.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
+   *   The entity type definition.
+   * @param \Drupal\Core\Entity\EntityManagerInterface $entity_manager
+   *   The entity manager service.
+   * @param \Drupal\Core\Language\LanguageManagerInterface $language_manager
+   *   The language manager.
+   * @param \Drupal\commerce_product\ProductVariationFieldRenderer $product_variation_field_renderer
+   *   The product variation field renderer
+   */
+  public function __construct(EntityTypeInterface $entity_type, EntityManagerInterface $entity_manager, LanguageManagerInterface $language_manager, ProductVariationFieldRenderer $product_variation_field_renderer) {
+    parent::__construct($entity_type, $entity_manager, $language_manager);
+    $this->productVariationFieldRenderer = $product_variation_field_renderer;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function createInstance(ContainerInterface $container, EntityTypeInterface $entity_type) {
+    return new static(
+      $entity_type,
+      $container->get('entity.manager'),
+      $container->get('language_manager'),
+      $container->get('commerce_product.variation_field_renderer')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildMultiple(array $build_list) {
+    // Build the view modes and display objects.
+    $view_modes = array();
+    $entity_type_key = "#{$this->entityTypeId}";
+    $view_hook = "{$this->entityTypeId}_view";
+
+    // Find the keys for the ContentEntities in the build; Store entities for
+    // rendering by view_mode.
+    $children = Element::children($build_list);
+    foreach ($children as $key) {
+      if (isset($build_list[$key][$entity_type_key])) {
+        $entity = $build_list[$key][$entity_type_key];
+        if ($entity instanceof FieldableEntityInterface) {
+          $view_modes[$build_list[$key]['#view_mode']][$key] = $entity;
+        }
+      }
+    }
+
+    // Build content for the displays represented by the entities.
+    foreach ($view_modes as $view_mode => $view_mode_entities) {
+      $displays = EntityViewDisplay::collectRenderDisplays($view_mode_entities, $view_mode);
+      $this->buildComponents($build_list, $view_mode_entities, $displays, $view_mode);
+      foreach (array_keys($view_mode_entities) as $key) {
+        // Allow for alterations while building, before rendering.
+        $entity = $build_list[$key][$entity_type_key];
+        $display = $displays[$entity->bundle()];
+
+        // Add product variation fields to the render array
+        $this->addProductVariationFields($build_list[$key], $entity, $view_mode, $key);
```

Comment:
So we have method invoked here, adds it to normal build process before invoking `entity_view` hooks.